File: ruby/robot-name/robot_name.rb
class Robot
  @@name_registry = {}

  def name
    @name ||= random_name
  end

  def random_name
    begin
      name = (('A'..'Z').to_a.sample(2) + (0..9).to_a.sample(3)).join('')
    end while (@@name_registry.key? name)
    @@name_registry[name] = 1
    name
  end

  def reset
    @@name_registry.delete(name)
    @name = nil
  end
end

module BookKeeping
  VERSION = 2
end

File: javascript/robot-name/robot-name.js
var usedNames = {};

Array.prototype.sample = function(size) {
  var sample = [];
  for (var i = 0; i < size; i++) {
    sample.push(this[Math.floor(Math.random() * this.length)]);
  }
  return sample;
}

var Robot = function() {

  var LETTERS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('');
  var DIGITS = '0123456789'.split('');

  var make_name = function() {
    do {
      var newName =  LETTERS.sample(2).join('') + DIGITS.sample(3).join('');
    } while(usedNames[newName] != undefined)
    usedNames[newName] = true;
    return newName;
  };

  this.name = '';

  this.reset = function() {
    var oldName = this.name;
    this.name = make_name();
    // usedNames[oldName] = undefined;
  }

  this.reset();
};

module.exports = Robot;
